### PR TITLE
[API compatibility] max pool with index return mask dtype set to int64 -part

### PR DESCRIPTION
**Pull request by jinyouzhi**
### PR Category

User Experience

### PR Types

New features

### Description

当 MaxPool*D 类函数参数 **return_mask=True** 时，第二个返回值 **mask** 的 Tensor **dtype 从 INT32 改为 INT64 ** 与pytorch对齐

涉及 API：
paddle.nn.functional.max_pool1d
paddle.nn.functional.max_pool2d
paddle.nn.functional.max_pool3d
paddle.nn.MaxPool1D
paddle.nn.MaxPool2D
paddle.nn.MaxPool3D
paddle.nn.functional.adaptive_max_pool1d
paddle.nn.functional.adaptive_max_pool2d
paddle.nn.functional.adaptive_max_pool3d
paddle.nn.AdaptiveMaxPool1D
paddle.nn.AdaptiveMaxPool2D
paddle.nn.AdaptiveMaxPool3D

例子：
```python
>>> import paddle.nn as nn
>>> input = paddle.uniform([1, 2, 3, 32, 32], dtype="float32", min=-1, max=1)

>>> MaxPool3D = nn.MaxPool3D(kernel_size=2, stride=2, padding=0, return_mask=True)
>>> output, max_indices = MaxPool3D(input)
>>> print(max_indices.dtype)
paddle.int64 # 原为 paddle.int32
```

 **XPU 暂时不修改**  （需要修改xpu sdk kernel）

### 是否引起精度变化

否

### CODE REVIEW

**Comment by jinyouzhi** [comment]
MaxPool(AdapativeMaxPoolxD/MaxPoolxD/FracionalMaxPoolxD) 系列API的 return_mask 参数跟 torch 还有diff：
* 修改mask返回dtype int32 成 int64 （这个pr）
* 参数名 return_mask 加上return_indices alias （后续
想请问一下这个修改计划，有concern吗？@zhwesky2010

**Comment by zhwesky2010** [comment]
@jinyouzhi 
1. XPU的问题是因为什么原因？
2. 这个修改可能会有一定的不兼容性风险，先看CI能不能都过，CI都能过可以先尝试合入，如果出现了后向不兼容问题，就只能暂且保持这种不一致了。
3. 这个需要加单测测试吗？
4. PR描述里，把会导致mask类型变化的所有API都列举清楚吧。有哪些API会变化，输出的哪个Tensor会变化写清楚。方便后面查问题。

**Comment by jinyouzhi** [comment]
> @jinyouzhi
> 
> 1. XPU的问题是因为什么原因？

XPU kernel 函数签名这个index_data签名是int32的，要改sdk或者加个转换也行，测了下不改能跑过ci就保持原状了。 
image


> 2. 这个修改可能会有一定的不兼容性风险，先看CI能不能都过，CI都能过可以先尝试合入，如果出现了后向不兼容问题，就只能暂且保持这种不一致了。

嗯是的；

> 3. 这个需要加单测测试吗？

单测改了对应mask的类型，兼容性的话，因为这个是个输出的类型，潜在的风险是对下游的影响，没太想到怎么测，可能组一个网测一下。

> 4. PR描述里，把会导致mask类型变化的所有API都列举清楚吧。有哪些API会变化，输出的哪个Tensor会变化写清楚。方便后面查问题。

ok。

**Comment by zhwesky2010** [comment]
看下CI没过的原因

**Comment by jinyouzhi** [comment]
> 看下CI没过的原因

ok，可以先看一下 https://github.com/PaddlePaddle/Paddle/pull/77655

**Comment by zhwesky2010** [comment]
@jinyouzhi 这个CI能通过吗，这两个PR最好一起合，避免有的返回int32，而有的返回int64。
如果这个PR过不了CI，建议另一个PR也不要合入。

**Comment by jinyouzhi** [comment]
> @jinyouzhi 这个CI能通过吗，这两个PR最好一起合，避免有的返回int32，而有的返回int64。
> 
> 如果这个PR过不了CI，建议另一个PR也不要合入。

ok明白了（本来考虑frac那个先看看情况再合入这个）我再re-run一下，我看了下coverage test cpp代码应该是maxpoolv2那个没有调用路径跑不到，python两行是静态图跑不到。

**Comment by jinyouzhi** [comment]
> 看下CI没过的原因

CI没过原因：
1 Linux-DCU 编不过 MaxPoolV2 算子 int64 改动，拿掉了MaxPoolV2后（内部用的cudnn算子，没接到F.max_pool/MaxPool）后过了
2 coverage 有两行静态的测不到
@zhwesky2010

**Comment by zhwesky2010** [comment]
@jinyouzhi 这两个PR暂时先不合入了，年前引入了不兼容风险，会不太可控。年后再看看。dtype为int32/64这个对不齐还好，差异比较细微，paconvert上可以设置check_dtype=False

**Comment by jinyouzhi** [comment]
> @jinyouzhi 这两个PR暂时先不合入了，年前引入了不兼容风险，会不太可控。年后再看看。

OK，有道理
